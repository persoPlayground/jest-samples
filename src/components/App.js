import React, { Component } from 'react';
import { Button } from 'react-bootstrap';
import Gift from './Gift';
import {getMaxId} from '../helpers';

class App extends Component {
state = {
    gifts: []
}

addGift = () => {
    const { gifts } = this.state;
    const ids = gifts.map(gift => gift.id );
    const maxId = getMaxId(ids);

    gifts.push({ id: maxId + 1 });
    this.setState({ gifts });

}
removeGift = id => {
    const gifts = this.state.gifts.filter(gift => gift.id !== id);
    this.setState({ gifts });
}
render(){
    return (
        <div>
            <h3>G-Giver</h3>
             <div className="gift-list">
             {
                 this.state.gifts.map( gift => {
                     return <Gift
                      key={gift.id}
                      gift = {gift}
                      removeGift={this.removeGift}
                      />
                 })
             }
             </div>
            <Button className="btn-add" onClick={this.addGift} >Add</Button>
           
        </div>    
    )
}
}

export default App;