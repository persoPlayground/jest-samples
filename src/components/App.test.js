import React from 'react';
import { shallow } from 'enzyme';
import App from './App';


describe('App', () => {

    const app = shallow(<App />);
    it('renders correctly', () => {
        expect(app).toMatchSnapshot();
    });
    
    it('initilizes the `state` as an empty list of gifts', () => {
        expect(app.state().gifts).toEqual([]);
    });
    

    describe('When clicking the Add button', () => {
        const id = 1;
        beforeEach(() => {
            app.find('.btn-add').simulate('click');
        });

        afterEach(() => {
            app.setState({ gifts: []});
        });
        it('adds a new gifts to `state`', () => {
            expect(app.state().gifts).toEqual([{ id}]);
        });
        
        it('adds a new gift to the rendred list', () => {
            expect(app.find('.gift-list').children().length).toEqual(1);
        });

        it('creates a gift component', () => {

            expect(app.find('Gift').exists()).toBe(true);
        });

        describe('when user wants to remove the last added gift', () => {
          beforeEach(() => {
            app.instance().removeGift(id);
          });

          it('removes the added gift', () =>{
            expect(app.state().gifts).toEqual([]);
          });

        });

    });
   
});
