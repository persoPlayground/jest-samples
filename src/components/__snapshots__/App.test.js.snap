// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          G-Giver
</h3>,
        <div
          className="gift-list"
>
          
</div>,
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          className="btn-add"
          disabled={false}
          onClick={[Function]}
>
          Add
</Button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "G-Giver",
        },
        "ref": null,
        "rendered": "G-Giver",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "gift-list",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "active": false,
          "block": false,
          "bsClass": "btn",
          "bsStyle": "default",
          "children": "Add",
          "className": "btn-add",
          "disabled": false,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Add",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            G-Giver
</h3>,
          <div
            className="gift-list"
>
            
</div>,
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn-add"
            disabled={false}
            onClick={[Function]}
>
            Add
</Button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "G-Giver",
          },
          "ref": null,
          "rendered": "G-Giver",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "gift-list",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": false,
            "block": false,
            "bsClass": "btn",
            "bsStyle": "default",
            "children": "Add",
            "className": "btn-add",
            "disabled": false,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Add",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
